### PR TITLE
added tesetcases for fetching patch files, + fix them!

### PR DESCRIPTION
**Pull request by JensTimmerman**
* patches with an explicit level 0 had the level ignored
* patches with a boolean as second argument had that boolean interpreted as an int

### CODE REVIEW

File: easybuild/framework/easyblock.py [hunk 1]
```diff
@@ -304,10 +304,12 @@ def fetch_patches(self, list_of_patches, extension=False, checksums=None):
             level = None
             if isinstance(patch_entry, (list, tuple)):
                 if not len(patch_entry) == 2:
-                    self.log.error("Unknown patch specification '%s', only two-element lists/tuples are supported!" % patch_entry)
+                    self.log.error("Unknown patch specification '%s', only two-element lists/tuples are supported!",
+                                   str(patch_entry))
                 pf = patch_entry[0]
 
-                if isinstance(patch_entry[1], int):
+                if type(patch_entry[1]) == int:  # int and only int is allowed here, we are parsing a config file, not
```
Comment (boegel) [inline easybuild/framework/easyblock.py]: how is this different from the `isinstance` call?

`isinstance` is the right way to check types imho

File: easybuild/framework/easyblock.py [hunk 2]
```diff
@@ -315,7 +317,10 @@ def fetch_patches(self, list_of_patches, extension=False, checksums=None):
                         copy_file = True
                     suff = patch_entry[1]
                 else:
-                    self.log.error("Wrong patch specification '%s', only int and string are supported as second element!" % patch_entry)
+                    self.log.error(
+                        "Wrong patch specification '%s', only int and string are supported as second element!",
```
Comment (boegel) [inline easybuild/framework/easyblock.py]: please collapse with line above, cfr. line 307?

if you need to, shorten the message (e.g. using "int/string", "2nd")

File: easybuild/framework/easyblock.py [hunk 3]
```diff
@@ -332,7 +337,7 @@ def fetch_patches(self, list_of_patches, extension=False, checksums=None):
                         patchspec['copy'] = suff
                     else:
                         patchspec['sourcepath'] = suff
-                if level:
+                if level is not None:
```
Comment (boegel) [inline easybuild/framework/easyblock.py]: how did you hit this exactly? is this case covered by the tests you've added now?

File: test/framework/easyblock.py [hunk 4]
```diff
@@ -407,6 +407,31 @@ def test_get_easyblock_instance(self):
...
 
+    def test_patchlevel(self):
+        """Test the parsing of the fetch_patches function."""
+        # adjust PYTHONPATH such that test easyblocks are found
+        testdir = os.path.abspath(os.path.dirname(__file__))
+        ec = process_easyconfig(os.path.join(testdir, 'easyconfigs', 'toy-0.0-patches.eb'))[0]
+        eb = get_easyblock_instance(ec)
+
+        patches = [
+            ('toy-0.0_level0_2.patch',0),  # should also be level 0 (not None)
```
Comment (boegel) [inline test/framework/easyblock.py]: style: space after `,`

File: test/framework/easyblock.py [hunk 5]
```diff
@@ -407,6 +407,31 @@ def test_get_easyblock_instance(self):
...
+            ('toy-0.0_level4.patch',4),   # should be level4
```
Comment (boegel) [inline test/framework/easyblock.py]: space after `,`

File: test/framework/easyblock.py [hunk 6]
```diff
@@ -407,6 +407,31 @@ def test_get_easyblock_instance(self):
...
+        ]
+        sandbox_sources = os.path.join(testdir, 'sandbox', 'sources')
+        init_config(args=["--sourcepath=%s" % sandbox_sources])
```
Comment (boegel) [inline test/framework/easyblock.py]: this is already done via `setUp` in `EnhancedTestCase`, so remove these two lines

File: test/framework/easyblock.py [hunk 7]
```diff
@@ -407,6 +407,31 @@ def test_get_easyblock_instance(self):
...
+        #check if patch levels are parsed correctly
```
Comment (boegel) [inline test/framework/easyblock.py]: `# check` (space)

File: test/framework/easyblock.py [hunk 8]
```diff
@@ -407,6 +407,31 @@ def test_get_easyblock_instance(self):
...
+        eb.fetch_patches(patches)
+
+        self.assertEquals(eb.patches[0]['level'], 0)
+        self.assertEquals(eb.patches[1]['level'], 4)
```
Comment (boegel) [inline test/framework/easyblock.py]: 4 patches are listed in the easyconfig, but you're checking only two?

the problem is that you're both listing patches in the easyconfig, and when calling `fetch_patches`

the `fetch_patches` code is a bit weird, it should be using `list_of_patches=None`, and if it's `None`, use `self.cfg['patches']`, i.e. the ones listed in the easyconfig

File: test/framework/easyconfigs/toy-0.0-patches.eb [hunk 9]
```diff
@@ -0,0 +1,32 @@
+name = 'toy'
```
Comment (boegel) [inline test/framework/easyconfigs/toy-0.0-patches.eb]: I don't think there's a need to add another toy easyconfig. `toy-0.0.eb` also contains a patch, so use that to test `fetch_patches` with the default `None`, so `self.cfg['patches'] is used

the other cases can be tested by supplying a list of patches to `fetch_patches`

File: easybuild/framework/easyblock.py (hunk 1 above)
Comment (JensTimmerman) [inline easybuild/framework/easyblock.py]: usually it is the right way to do it, since you want to have duck typing, 
however, when parsing config files you want to be rather strickt, and no duck typing.
The bug arising in this case is that bool is an instance of int, but you really want to discern between  an int and a bool, see also the inline comment.

File: easybuild/framework/easyblock.py (hunk 2 above)
Comment (JensTimmerman) [inline easybuild/framework/easyblock.py]: no, this is perfectly valid.

File: easybuild/framework/easyblock.py (hunk 3 above)
Comment (JensTimmerman) [inline easybuild/framework/easyblock.py]: yes, try the test cases with and without this line if you're curious, 

when setting patch level to 0 this if returns false, but you actually want to set '0' not ignore it.
 (later on the guesser would guess level 4, but I explicitly put level 0, so it shouldn't have guessed.
Maybe the guesser could also be fixed, but the patch file was rather deep, and there was a weird directory structure, so I don't blame it for not guessing correct, that's a different issue altogether)

File: test/framework/easyblock.py (hunk 8 above)
Comment (JensTimmerman) [inline test/framework/easyblock.py]: no, I'm checking the 3rd one later, it should raise an error.
and the first one was already there.

For these test cases it doesn't really matter what patch files are in the easyconfig, because of the weird code I'm actually never using the ones in the easyconfig, so the problem you are describing is not a problem. (if it were executed, it should throw an error on the 4th one)

I just needed an easyconfig, so I made a new one, I can remove it again if you like.

File: test/framework/easyconfigs/toy-0.0-patches.eb (hunk 9 above)
Comment (JensTimmerman) [inline test/framework/easyconfigs/toy-0.0-patches.eb]: actually I don't wat to test with None, I'm testing easyblock code here, not easyconfig code.

File: easybuild/framework/easyblock.py (hunk 3 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: It's only guessing, I don't think there's a way to fix it. Thanks for clarifying that setting the patch level to `0` revealed this issue.

File: easybuild/framework/easyblock.py (hunk 1 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: what, `bool` is an instance of `int`? eek!!

in that case, ok, this makes sense, thanks for clarifying it

File: easybuild/framework/easyblock.py (hunk 2 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: it's stupid, we already have enough line of code to scroll around in, there's absolutely no reason why this should be spread across 4 lines of code, it's madness

**Comment by boegel** [comment]
merged in via #1139
